### PR TITLE
Upgrade Emscripten targets to use upstream LLVM backend

### PR DESCRIPTION
**Pull request by tlively**
- Compatible with Emscripten 1.38.46-upstream or later upstream.
 - Refactors the Emscripten target spec to share code with other wasm
   targets.
 - Replaces the old incorrect wasm32 C call ABI with the correct one,
   preserving the old one as wasm32_bindgen_compat for wasm-bindgen
   compatibility.
 - Updates the varargs ABI used by Emscripten and deletes the old one.
 - Removes the obsolete wasm32-experimental-emscripten target.
 - Uses EMCC_CFLAGS on CI to avoid the timeout problems with #63649.

r? @alexcrichton

### CODE REVIEW

**Comment by alexcrichton** [comment]
@bors: delegate+

Looks good to me! [As discussed here](https://github.com/rust-lang/rust/pull/65151#issuecomment-539677656) looks like we're going to disable the asmjs tests here but enable the wasm32-unknown-emscripten ones instead. These should execute much more quickly and continue to be within our time limits on CI

@tlively I'm assuming this is basically https://github.com/rust-lang/rust/pull/63649 plus that change for the final pass, so once CI comes back green feel free to r=me!

**Comment by cup** [comment]
https://github.com/rust-lang/rust/issues/65296

File: src/liballoc/tests/str.rs.rej [hunk 1]
```diff
@@ -0,0 +1,10 @@
+diff a/src/liballoc/tests/str.rs b/src/liballoc/tests/str.rs	(rejected hunks)
```
Comment (mominul) [inline src/liballoc/tests/str.rs.rej]: Why this file is included?

Comment (tlively) [reply src/liballoc/tests/str.rs.rej]: It should not be! Thanks for the catch.

**Comment by tlively** [comment]
I have no idea why all those tests started failing suddenly, and I can't reproduce it locally. I'll just rebase on master and push again and see if it goes away :confused:

**Comment by tlively** [comment]
Well I guess I'll just wait until docker hub is back up :upside_down_face:

**Comment by tlively** [comment]
@bors r+

@alexcrichton No major tweaks since you last accepted this. The only possibly sketchy thing is the use of NO_CHANGE_USER in the asmjs and wasm32 docker files in https://github.com/rust-lang/rust/pull/65251/commits/4b26d9c142efcbd175dad36f7cb4ebf9e682154e. But hopefully that's less sketchy than the shenanigans committed upon the emsdk install before that commit.

**Comment by alexcrichton** [comment]
Thanks again @tlively for helping to push this over the finish line!

**Comment by XAMPPRocky** [comment]
Marking as relnotes for myself in the future, as the previous PR was marked as such and I assume it's still desired to include in the release notes of the release.

File: src/test/codegen/c-variadic.rs [hunk 2]
```diff
@@ -1,3 +1,4 @@
+// ignore-emscripten compiled with panic=abort by default
```
Comment (RalfJung) [inline src/test/codegen/c-variadic.rs:1]: The test is about variadics, not panics, so I do not understand this comment.

(I realize it's been a while, but I just coincidentally noticed this comment and traced it back to where it was added...)

Comment (tlively) [reply src/test/codegen/c-variadic.rs:1]: My guess is that compiling with `panic=abort` changes the codegen in such a way that some of the CHECK lines no longer pass (particularly down in `test_c_variadic_call`, which does not have an explicit panic annotation).

**Pull request merged**